docs/discussion/lwip_ipv6.md

```diff
@@ -0,0 +1,139 @@
+lwip layer. This seems like a good model for CHIP going forward.
+
+Additionally, we should ensure that CHIP uses non-blocking DNS APIs.
+
+### Recommendation:
+
+-   bug fix for DHCPv6 to avoid it setting bad addresses.
+    -   note - fixed in
+        https://git.savannah.nongnu.org/cgit/lwip.git/commit/?id=941300c21c45a4dbf1c074b29a9ca3c88c9f6553,
+        but not yet released as a part of an official release.
+-   Create a patch to add hooks to the SetDns and GetDns functions so logic for
+    selecting the DNS server can be moved into the manager layer (possibly
+    already in CHIP, but not present in other middleware)
+
+## Address Scopes
+
+Link local addresses are less common on IPv4, which normally rely on NAT at the
+router to do address translation. CHIP mandates the use of IPv6 link local
+addresses for communication to nodes on the same network (wifi or thread). When
+there is more than one netif in the system (ex. loopback, softAP, STA), the link
+local address needs more information to determine which link the address is
+local to. This is normally added as the link local scope and can be seen on
+addresses ex. FE80::xxxx:xxxx:xxxx:xxxx%<scope>, where the <scope> identifies
+the netif (something like %wlan0 or %eno1 etc.).
+
+Without this indicator, the link local address can only be resolved if there is
+one netif. LwIP will also allow a direct address match to the netif source
+address, but this does not scale well at all and is VERY racy. LwIP also
+supports output to a specific netif, but this is not brought up to the socket
+layer.
+
+Upstream LwIP has support for IPv6 address scopes, but only as an option.
+However, the code to support this is not present in the CHIP LwIP codebase.
+Other platform versions assume this option is not present (ex. M5 has an
+assertion on ip address sizes that disallow the use of a scope tag).
+
+### Recommendation:
+
+-   Update CHIP lwip version for scopes support, ensure CHIP code works with
+    scopes on our various platforms OR alternate: bring netif sendto up through
+    the api / sockets layers
+-   Audit CHIP code to ensure LL addresses are properly scoped to their netif in
+    all areas (DNS returned addresses especially)
+
+## Duplicate address detection
+
+The DAD in LwIP is actually implemented correctly right now, but there are
+routers that incorrectly implement multicast for IPv6 and send packets back to
+the sender. This triggers the LwIP DAD because it doesn’t check the source. In
+our previous products, we fixed this as the wifi layer as a filter, but it’s
+easy enough to add the fix into the LwIP layer. This would help implementers so
+they don’t all have to debug the same issues. Recommendation:
+
+-   Create an LwIP patch to check NS/NA packets for source and discard if they
+    originate from the same device. Upstream and add to CHIP.
+
+## IPv6 Ping
+
+LwIP will automatically respond to pings, but has no built-in way to send them.
```

Comment:
Not sure why we need to implement ICMPv6 ping for Matter, although it is a good idea. It is not required by the spec to send ICMPv6 pings.

Reply:
We don't need it for chip, but given that this is probably one of the first times a lot of devs will be working exclusively with IPv6 - ping can be a really helpful tool. I'll add a comment clarifying that this is something we should consider as a tool for development, not as a requirement of the spec.